streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreTest.java

```diff
@@ -523,6 +537,63 @@ public void shouldRestoreMultipleBatches() {
         verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 15);
     }
 
+    @Test
+    public void shouldNotRestoreExpired() {
+        final List<DataRecord> records = new ArrayList<>();
+        records.add(new DataRecord("k", "v", HISTORY_RETENTION + 10));
+        records.add(new DataRecord("k1", "v1", HISTORY_RETENTION + 10 - GRACE_PERIOD)); // grace period has not elapsed
+        records.add(new DataRecord("k2", "v2", HISTORY_RETENTION + 9 - GRACE_PERIOD)); // grace period has elapsed, so this record should not be restored
```

Comment:
Cf comment above. The question seems to be "when" the original `put()` happened with regard to stream-time?

Reply:
That's correct. This test case uses the same data as `shouldNotPutExpired()` above. This third record is expired even during normal put operations.

Reply:
Thx. After you pointed out the "store hierarchy" in your above reply and that the record would go into the changelog, the test makes sense.